build_tools/docker/manage_images.py

```diff
@@ -130,144 +129,133 @@ def parse_arguments():
+  else:
     print(f'{" ".join(grep_command)} '
-          f'failed with exit code {grep_process.returncode}')
-    sys.exit(grep_process.returncode)
-  if grep_process.returncode == 1:
-    print(f'Found no references to {image_name}')
-    return
-
-  files = grep_process.stdout.split()
-  print(f'Updating references in {len(files)} files: {files}')
-  for line in fileinput.input(files=files, inplace=(not dry_run)):
-    print(re.sub(f'{image_name}@sha256:[a-zA-Z0-9]+', f'{image_name}@{digest}',
-                 line),
-          end='')
+          f'failed with exit code {completed_process.returncode}')
+    sys.exit(completed_process.returncode)
 
 
 if __name__ == '__main__':
   args = parse_arguments()
 
-  # Ensure the user has the correct authorization if they try to push to GCR.
   if args.push:
-    if stream_command(['which', 'gcloud']) != 0:
-      print('gcloud not found.'
-            ' See https://cloud.google.com/sdk/install for installation.')
-      sys.exit(1)
-    check_stream_command(['gcloud', 'auth', 'configure-docker'],
-                         dry_run=args.dry_run)
+    # Ensure the user has the correct authorization if they try to push to GCR.
+    utils.check_gcloud_auth(dry_run=args.dry_run)
 
   images_to_process = get_ordered_images_to_process(args.images)
   print(f'Also processing dependent images. Will process: {images_to_process}')
 
   for image in images_to_process:
     print(f'Processing image {image}')
-    image_name = posixpath.join(IREE_GCR_URL, image)
-    image_tag = f'{image_name}:{args.tag}'
+    image_url = posixpath.join(IREE_GCR_URL, image)
+    tagged_image_url = f'{image_url}:{args.tag}'
     image_path = os.path.join(DOCKER_DIR, image)
 
     if args.pull:
-      check_stream_command(['docker', 'pull', image_tag], dry_run=args.dry_run)
+      utils.run_command(['docker', 'pull', tagged_image_url], args.dry_run)
 
     if args.build:
-      check_stream_command(['docker', 'build', '--tag', image_tag, image_path],
-                           dry_run=args.dry_run)
+      utils.run_command(
+          ['docker', 'build', '--tag', tagged_image_url, image_path],
+          args.dry_run)
 
     if args.push:
-      check_stream_command(['docker', 'push', image_tag], dry_run=args.dry_run)
+      if args.tag == 'prod':
+        raise ValueError('This script cannot be used to push to :prod. Use '
+                         'build_tools/docker/manage_prod.py instead.')
+      utils.run_command(['docker', 'push', tagged_image_url], args.dry_run)
 
     if args.update_references:
-      digest = get_repo_digest(image_tag)
+      digest = get_repo_digest(tagged_image_url)
+
+      # Check that the image is in 'prod_digests.txt' and append it to the list
+      # in the file if it isn't. We know that the GCR digest exists at this
+      # point because 'get_repo_digest' confirms that the image has been pushed.
+      with open(utils.PROD_DIGESTS, 'r') as f:
+        in_prod_digests = image_url in f.read()
```

Comment:
One URL could be a prefix of another. In fact we already have that case. This will lead to false positives. I think if you anchor with "@" that will solve the issue

```suggestion
        in_prod_digests = f'{image_url}@' in f.read()
```

Reply:
Good catch.